tests/test_runtime_errors.py

```diff
@@ -108,43 +107,6 @@ def test_unable_to_reconstruct_message_pythonpath(monkeypatch: "MonkeyPatch") ->
     assert re.search(message, str(output.stderr))
 
 
-def test_unable_to_parse() -> None:
-    """Check error message shows if we're unable to parse notebook."""
-    path = Path("tests") / "data/invalid_notebook.ipynb"
-    path.write_text("foo")
-    message = f"Error parsing {str(path)}"
-    with pytest.raises(RuntimeError) as excinfo:
-        main(["flake8", str(path), "--nbqa-mutate"])
-    path.unlink()
-    assert message in str(excinfo.value)
```

Comment:
On Windows, during CI, this gives
```python
================================== FAILURES ===================================
____________________________ test_unable_to_parse _____________________________

cli_args = <nbqa.cmdline.CLIArgs object at 0x000002860FE4BBE0>
configs = <nbqa.config.config.Configs object at 0x000002860FE4B5C0>

    def _run_on_one_root_dir(  # pylint: disable=R0912
        cli_args: CLIArgs, configs: Configs
    ) -> int:
        """
        Run third-party tool on a single notebook or directory.
    
        Parameters
        ----------
        cli_args
            Commanline arguments passed to nbqa.
        configs
            Configuration passed to nbqa from commandline or via a config file
    
        Returns
        -------
        int
            Output code from third-party tool.
    
        Raises
        ------
        RuntimeError
            If unable to parse or reconstruct notebook.
        SystemExit
            If third-party tool would've reformatted notebook but ``--nbqa-mutate``
            wasn't passed.
    
        Raises
        ------
        FileNotFoundError
            If notebook which doesn't exist is passed.
        """
        nb_to_py_mapping = {}
        for notebook in _get_all_notebooks(
            cli_args.root_dirs, configs.nbqa_files, configs.nbqa_exclude
        ):
            if not notebook.exists():
                raise FileNotFoundError(
                    f"{BOLD}No such file or directory: {str(notebook)}{RESET}"
                )
    
            nb_to_py_mapping[notebook] = tempfile.mkstemp(
                dir=str(notebook.parent),
                prefix=notebook.stem,
                suffix=".py",
            )
    
        try:  # pylint disable=R0912
    
            if not nb_to_py_mapping:
                sys.stderr.write(
                    "No .ipynb notebooks found in given directories: "
                    f"{' '.join(i for i in cli_args.root_dirs if Path(i).is_dir())}\n"
                )
                return 0
    
            nb_info_mapping: MutableMapping[Path, NotebookInfo] = {}
    
            for notebook, (file_descriptor, _) in nb_to_py_mapping.items():
                try:
                    nb_info_mapping[notebook] = save_source.main(
                        notebook,
                        file_descriptor,
                        configs.nbqa_process_cells,
>                       cli_args.command,
                    )

nbqa\__main__.py:361: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

notebook = WindowsPath('tests/data/invalid_notebook.ipynb')
file_descriptor = 12, process_cells = [], command = 'flake8'

    def main(  # pylint: disable=R0914
        notebook: "Path",
        file_descriptor: int,
        process_cells: Sequence[str],
        command: str,
    ) -> NotebookInfo:
        """
        Extract code cells from notebook and save them in temporary Python file.
    
        Parameters
        ----------
        notebook
            Jupyter Notebook third-party tool is being run against.
        process_cells
            Extra cells which nbqa should process.
        command
            The third-party tool being run.
    
        Returns
        -------
        NotebookInfo
    
        """
>       cells = json.loads(notebook.read_text(encoding="utf-8"))["cells"]

nbqa\save_source.py:329: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

s = 'foo', encoding = None, cls = None, object_hook = None, parse_float = None
parse_int = None, parse_constant = None, object_pairs_hook = None, kw = {}

    def loads(s, *, encoding=None, cls=None, object_hook=None, parse_float=None,
            parse_int=None, parse_constant=None, object_pairs_hook=None, **kw):
        """Deserialize ``s`` (a ``str``, ``bytes`` or ``bytearray`` instance
        containing a JSON document) to a Python object.
    
        ``object_hook`` is an optional function that will be called with the
        result of any object literal decode (a ``dict``). The return value of
        ``object_hook`` will be used instead of the ``dict``. This feature
        can be used to implement custom decoders (e.g. JSON-RPC class hinting).
    
        ``object_pairs_hook`` is an optional function that will be called with the
        result of any object literal decoded with an ordered list of pairs.  The
        return value of ``object_pairs_hook`` will be used instead of the ``dict``.
        This feature can be used to implement custom decoders that rely on the
        order that the key and value pairs are decoded (for example,
        collections.OrderedDict will remember the order of insertion). If
        ``object_hook`` is also defined, the ``object_pairs_hook`` takes priority.
    
        ``parse_float``, if specified, will be called with the string
        of every JSON float to be decoded. By default this is equivalent to
        float(num_str). This can be used to use another datatype or parser
        for JSON floats (e.g. decimal.Decimal).
    
        ``parse_int``, if specified, will be called with the string
        of every JSON int to be decoded. By default this is equivalent to
        int(num_str). This can be used to use another datatype or parser
        for JSON integers (e.g. float).
    
        ``parse_constant``, if specified, will be called with one of the
        following strings: -Infinity, Infinity, NaN.
        This can be used to raise an exception if invalid JSON numbers
        are encountered.
    
        To use a custom ``JSONDecoder`` subclass, specify it with the ``cls``
        kwarg; otherwise ``JSONDecoder`` is used.
    
        The ``encoding`` argument is ignored and deprecated.
    
        """
        if isinstance(s, str):
            if s.startswith('\ufeff'):
                raise JSONDecodeError("Unexpected UTF-8 BOM (decode using utf-8-sig)",
                                      s, 0)
        else:
            if not isinstance(s, (bytes, bytearray)):
                raise TypeError('the JSON object must be str, bytes or bytearray, '
                                'not {!r}'.format(s.__class__.__name__))
            s = s.decode(detect_encoding(s), 'surrogatepass')
    
        if (cls is None and object_hook is None and
                parse_int is None and parse_float is None and
                parse_constant is None and object_pairs_hook is None and not kw):
>           return _default_decoder.decode(s)

c:\hostedtoolcache\windows\python\3.6.8\x64\lib\json\__init__.py:354: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <json.decoder.JSONDecoder object at 0x000002860D45F5F8>, s = 'foo'
_w = <built-in method match of _sre.SRE_Pattern object at 0x000002860D4531B0>

    def decode(self, s, _w=WHITESPACE.match):
        """Return the Python representation of ``s`` (a ``str`` instance
        containing a JSON document).
    
        """
>       obj, end = self.raw_decode(s, idx=_w(s, 0).end())

c:\hostedtoolcache\windows\python\3.6.8\x64\lib\json\decoder.py:339: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <json.decoder.JSONDecoder object at 0x000002860D45F5F8>, s = 'foo'
idx = 0

    def raw_decode(self, s, idx=0):
        """Decode a JSON document from ``s`` (a ``str`` beginning with
        a JSON document) and return a 2-tuple of the Python
        representation and the index in ``s`` where the document ended.
    
        This can be used to decode a JSON document from a string that may
        have extraneous data at the end.
    
        """
        try:
            obj, end = self.scan_once(s, idx)
        except StopIteration as err:
>           raise JSONDecodeError("Expecting value", s, err.value) from None
E           json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

c:\hostedtoolcache\windows\python\3.6.8\x64\lib\json\decoder.py:357: JSONDecodeError

The above exception was the direct cause of the following exception:

cli_args = <nbqa.cmdline.CLIArgs object at 0x000002860FE4BBE0>
configs = <nbqa.config.config.Configs object at 0x000002860FE4B5C0>

    def _run_on_one_root_dir(  # pylint: disable=R0912
        cli_args: CLIArgs, configs: Configs
    ) -> int:
        """
        Run third-party tool on a single notebook or directory.
    
        Parameters
        ----------
        cli_args
            Commanline arguments passed to nbqa.
        configs
            Configuration passed to nbqa from commandline or via a config file
    
        Returns
        -------
        int
            Output code from third-party tool.
    
        Raises
        ------
        RuntimeError
            If unable to parse or reconstruct notebook.
        SystemExit
            If third-party tool would've reformatted notebook but ``--nbqa-mutate``
            wasn't passed.
    
        Raises
        ------
        FileNotFoundError
            If notebook which doesn't exist is passed.
        """
        nb_to_py_mapping = {}
        for notebook in _get_all_notebooks(
            cli_args.root_dirs, configs.nbqa_files, configs.nbqa_exclude
        ):
            if not notebook.exists():
                raise FileNotFoundError(
                    f"{BOLD}No such file or directory: {str(notebook)}{RESET}"
                )
    
            nb_to_py_mapping[notebook] = tempfile.mkstemp(
                dir=str(notebook.parent),
                prefix=notebook.stem,
                suffix=".py",
            )
    
        try:  # pylint disable=R0912
    
            if not nb_to_py_mapping:
                sys.stderr.write(
                    "No .ipynb notebooks found in given directories: "
                    f"{' '.join(i for i in cli_args.root_dirs if Path(i).is_dir())}\n"
                )
                return 0
    
            nb_info_mapping: MutableMapping[Path, NotebookInfo] = {}
    
            for notebook, (file_descriptor, _) in nb_to_py_mapping.items():
                try:
                    nb_info_mapping[notebook] = save_source.main(
                        notebook,
                        file_descriptor,
                        configs.nbqa_process_cells,
                        cli_args.command,
                    )
                except Exception as exc:  # pragma: nocover
                    raise RuntimeError(
                        BASE_ERROR_MESSAGE.format(f"Error parsing {str(notebook)}")
>                   ) from exc
E                   RuntimeError: Error parsing tests\data\invalid_notebook.ipynb
E                   Please report a bug at https://github.com/nbQA-dev/nbQA/issues 

nbqa\__main__.py:366: RuntimeError

During handling of the above exception, another exception occurred:

    def test_unable_to_parse() -> None:
        """Check error message shows if we're unable to parse notebook."""
        path = os.path.join("tests", "data", "invalid_notebook.ipynb")
        try:
            with open(path, "w") as handle:
                handle.write("foo")
            message = f"Error parsing {str(path)}"
            with pytest.raises(RuntimeError) as excinfo:
>               main(["flake8", str(path), "--nbqa-mutate"])

tests\test_runtime_errors.py:118: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
nbqa\__main__.py:488: in main
    output_code = _run_on_one_root_dir(cli_args, configs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cli_args = <nbqa.cmdline.CLIArgs object at 0x000002860FE4BBE0>
configs = <nbqa.config.config.Configs object at 0x000002860FE4B5C0>

    def _run_on_one_root_dir(  # pylint: disable=R0912
        cli_args: CLIArgs, configs: Configs
    ) -> int:
        """
        Run third-party tool on a single notebook or directory.
    
        Parameters
        ----------
        cli_args
            Commanline arguments passed to nbqa.
        configs
            Configuration passed to nbqa from commandline or via a config file
    
        Returns
        -------
        int
            Output code from third-party tool.
    
        Raises
        ------
        RuntimeError
            If unable to parse or reconstruct notebook.
        SystemExit
            If third-party tool would've reformatted notebook but ``--nbqa-mutate``
            wasn't passed.
    
        Raises
        ------
        FileNotFoundError
            If notebook which doesn't exist is passed.
        """
        nb_to_py_mapping = {}
        for notebook in _get_all_notebooks(
            cli_args.root_dirs, configs.nbqa_files, configs.nbqa_exclude
        ):
            if not notebook.exists():
                raise FileNotFoundError(
                    f"{BOLD}No such file or directory: {str(notebook)}{RESET}"
                )
    
            nb_to_py_mapping[notebook] = tempfile.mkstemp(
                dir=str(notebook.parent),
                prefix=notebook.stem,
                suffix=".py",
            )
    
        try:  # pylint disable=R0912
    
            if not nb_to_py_mapping:
                sys.stderr.write(
                    "No .ipynb notebooks found in given directories: "
                    f"{' '.join(i for i in cli_args.root_dirs if Path(i).is_dir())}\n"
                )
                return 0
    
            nb_info_mapping: MutableMapping[Path, NotebookInfo] = {}
    
            for notebook, (file_descriptor, _) in nb_to_py_mapping.items():
                try:
                    nb_info_mapping[notebook] = save_source.main(
                        notebook,
                        file_descriptor,
                        configs.nbqa_process_cells,
                        cli_args.command,
                    )
                except Exception as exc:  # pragma: nocover
                    raise RuntimeError(
                        BASE_ERROR_MESSAGE.format(f"Error parsing {str(notebook)}")
                    ) from exc
    
            out, err, output_code, mutated = _run_command(
                cli_args.command,
                configs.nbqa_addopts,
                [i[1] for i in nb_to_py_mapping.values()],
            )
    
            for notebook, (_, temp_python_file) in nb_to_py_mapping.items():
                out, err = _replace_temp_python_file_references_in_out_err(
                    temp_python_file, notebook, out, err
                )
                try:
                    out, err = map_python_line_to_nb_lines(
                        cli_args.command,
                        out,
                        err,
                        notebook,
                        nb_info_mapping[notebook].cell_mappings,
                    )
                except Exception as exc:  # pylint: disable=W0703  # pragma: nocover
                    msg = (
                        f"{repr(exc)} while parsing output "
                        f"from applying {cli_args.command} to {str(notebook)}"
                    )
                    sys.stderr.write(BASE_ERROR_MESSAGE.format(msg))
    
                if mutated:
                    if not configs.nbqa_mutate and not configs.nbqa_diff:
                        # pylint: disable=C0301
                        msg = dedent(
                            f"""\
                            {BOLD}Mutation detected, will not reformat! Please use the `--nbqa-mutate` flag, e.g.:{RESET}
    
                                nbqa {cli_args.command} notebook.ipynb --nbqa-mutate
    
                            or, to only preview changes, use the `--nbqa-diff` flag, e.g.:
    
                                nbqa {cli_args.command} notebook.ipynb --nbqa-diff
                            """
                        )
                        # pylint: enable=C0301
                        raise SystemExit(msg)
    
                    try:
                        REPLACE_FUNCTION[configs.nbqa_diff](
                            temp_python_file,
                            notebook,
                            nb_info_mapping[notebook],
                        )
                    except Exception as exc:
                        raise RuntimeError(
                            BASE_ERROR_MESSAGE.format(
                                f"Error reconstructing {str(notebook)}"
                            )
                        ) from exc
    
            sys.stdout.write(out)
            sys.stderr.write(err)
    
            if configs.nbqa_diff:
                if mutated:
                    sys.stdout.write(
                        "To apply these changes use `--nbqa-mutate` instead of `--nbqa-diff`\n"
                    )
                return output_code
    
        finally:
            for _, tmp_path in nb_to_py_mapping.values():
>               os.remove(tmp_path)
E               PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'D:\\a\\nbQA\\nbQA\\tests\\data\\invalid_notebook6lremlo8.py'

nbqa\__main__.py:435: PermissionError
=========================== short test summary info ===========================
FAILED tests/test_runtime_errors.py::test_unable_to_parse - PermissionError: ...
======================== 1 failed, 75 passed in 46.74s ========================
ERROR: InvocationError for command 'D:\a\nbQA\nbQA\.tox\py\Scripts\coverage.EXE' run -m pytest tests -vv -W error (exited with code 1)
___________________________________ summary ___________________________________
```

which seems like a real concern

Reply:
However, other tests pass on Windows, so...perhaps the issue is with the test?